### PR TITLE
Use Android project modifier for post-processing

### PR DESCRIPTION
**Pull request by aurimasc**
https://jira.unity3d.com/browse/PLAT-10985
Use AndroidProjectFilesModifier for build post-processing in Unity 6 for better compatibility with incremental build.

Changes:
- Building project now has to different paths, one for Unity 6 and other for older versions. This impacts how notification settings are applied to project (custom icons and android manifest file).
- Instead of having default manifest in .androidlib and patching it during post process now a file is generated. Should not have functional difference.

Focus for testing:
- Android only, unrelated to Android version nor device, only affects gradle project generation
- Unity 2021.3 and 2022.3 should behave the same, no need to test both; Unity 6 is different.
- Icons set in notifications settings need to end up in gradle project properly. I've only tested Main project that notifications with icons show up correctly.
- The generated manifest file in notifications sub-project should be *meaningfully* the same in Unity 6, older and the package before these changes. There might be text differences, but the items should be the same. I've tested this manually, but worth double checking as some setting might have been overlooked.

### CODE REVIEW

File: com.unity.mobile.notifications/Editor/AndroidNotificationPostProcessor.cs [hunk 1]
```diff
@@ -1,18 +1,202 @@
...
+                    if (useExact)
+                        sheduleExactAlarm.Attributes.MaxSdkVersion.Set(32);
+                }
+                if (useExact)
+                    manifest.AddUsesPermission("android.permission.USE_EXACT_ALARM");
+
+                // Battery optimizations must use "uses-permission-sdk-23", regular uses-permission does not work
+                if ((settings.ExactAlarm & AndroidExactSchedulingOption.AddRequestIgnoreBatteryOptimizationsPermission) != 0)
+                {
+                    // TODO: Missing AddUsesPermissionSdk23 function
```
[the remark this answers is not in the document]
Comment (aurimasc) [reply com.unity.mobile.notifications/Editor/AndroidNotificationPostProcessor.cs:172]: Hm, does not seem to be the case: https://docs.unity3d.com/6000.0/Documentation/ScriptReference/Unity.Android.Gradle.Manifest.Manifest.html

Strangely, we do have UsesPermissionSdk23List. Or does it mean one should call AddElement() on a list returned by that one?

File: com.unity.mobile.notifications/Editor/AndroidNotificationPostProcessor.cs [hunk 2]
```diff
@@ -1,18 +1,202 @@
...
+        }
+
+        private static T GetSetting<T>(List<NotificationSetting> settings, string key)
+        {
+            return (T)settings.Find(i => i.Key == key).Value;
+        }
+    }
+
+#if UNITY_2023_1_OR_NEWER
+    public class AndroidNotificationPostProcessor : AndroidProjectFilesModifier
```
[the remark this answers is not in the document]
Comment (aurimasc) [reply com.unity.mobile.notifications/Editor/AndroidNotificationPostProcessor.cs]: Removed.

File: com.unity.mobile.notifications/Editor/AndroidNotificationPostProcessor.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (todi1856) [reply com.unity.mobile.notifications/Editor/AndroidNotificationPostProcessor.cs:172]: it seems we forgot to adds docs, but should do you something like this
```
batterOptimizations.AddUsesPermissionSdk23("android.permission.REQUEST_IGNORE_BATTERY_OPTIMIZATIONS");
```
and that's it.

(Please validate that it does add it correctly)

File: com.unity.mobile.notifications/Editor/AndroidNotificationPostProcessor.cs [hunk 3]
```diff
@@ -1,18 +1,202 @@
...
+        }
+
+        private static void InjectReceivers(Manifest manifest)
+        {
+            var receiverkNotificationManager = new Receiver();
+            receiverkNotificationManager.Attributes.Name.Set("com.unity.androidnotifications.UnityNotificationManager");
+            receiverkNotificationManager.Attributes.Exported.Set(false);
+            manifest.Application.ReceiverList.AddElement(receiverkNotificationManager);
+
+            var receiverNotificationRestartOnBoot = new Receiver();
```
[the remark this answers is not in the document]
Comment (aurimasc) [reply com.unity.mobile.notifications/Editor/AndroidNotificationPostProcessor.cs]: Fixed.

File: com.unity.mobile.notifications/Editor/AndroidNotificationPostProcessor.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (aurimasc) [reply com.unity.mobile.notifications/Editor/AndroidNotificationPostProcessor.cs]: Yes, it works.
I've also changed defines to Unity 6.

**Review by todi1856 — APPROVED** [review]

**Review by IGuscin — CHANGES_REQUESTED** [review]
Found small issue on older Unity version (2022.3, issue exist in main as well, so not introduced by this PR) that entries are not removed from manifest when settings are changed. i.e. export project and check manifest when Schedule at exact time = nothing. Then do the same with Everything. And again with Nothing. Manifest will remain the same as with Everything.
Same issue is not reproducible on Unity 6, since the way manifest generated differs.
Otherwise looks good, manifest content is basically the same (package="com.unity.androidnotifications"> is now missing, but that's expected and order of elements is a bit different, but that doesn't impact functionality)

**Review by IGuscin — APPROVED** [review]

**Pull request merged**